src/main.rs

```diff
@@ -0,0 +1,574 @@
+	keys_manager: Arc<KeysManager>, payment_storage: PaymentInfoStorage, network: Network,
+) {
+	let mut pending_txs: HashMap<OutPoint, Transaction> = HashMap::new();
+	loop {
+		peer_manager.process_events();
+		let loop_channel_manager = channel_manager.clone();
+		let mut events = channel_manager.get_and_clear_pending_events();
+		events.append(&mut chain_monitor.get_and_clear_pending_events());
+		for event in events {
+			match event {
+				Event::FundingGenerationReady {
+					temporary_channel_id,
+					channel_value_satoshis,
+					output_script,
+					..
+				} => {
+					// Construct the raw transaction with one output, that is paid the amount of the
+					// channel.
+					let addr = WitnessProgram::from_scriptpubkey(
+						&output_script[..],
+						match network {
+							Network::Bitcoin => bitcoin_bech32::constants::Network::Bitcoin,
+							Network::Testnet => bitcoin_bech32::constants::Network::Testnet,
+							Network::Regtest => bitcoin_bech32::constants::Network::Regtest,
+							Network::Signet => panic!("Signet unsupported"),
+						},
+					)
+					.expect("Lightning funding tx should always be to a SegWit output")
+					.to_address();
+					let mut outputs = vec![HashMap::with_capacity(1)];
+					outputs[0].insert(addr, channel_value_satoshis as f64 / 100_000_000.0);
+					let raw_tx = bitcoind_client.create_raw_transaction(outputs).await;
+
+					// Have your wallet put the inputs into the transaction such that the output is
+					// satisfied.
+					let funded_tx = bitcoind_client.fund_raw_transaction(raw_tx).await;
+					let change_output_position = funded_tx.changepos;
+					assert!(change_output_position == 0 || change_output_position == 1);
+
+					// Sign the final funding transaction and broadcast it.
+					let signed_tx =
+						bitcoind_client.sign_raw_transaction_with_wallet(funded_tx.hex).await;
+					assert_eq!(signed_tx.complete, true);
+					let final_tx: Transaction =
+						encode::deserialize(&hex_utils::to_vec(&signed_tx.hex).unwrap()).unwrap();
+					let outpoint = OutPoint {
+						txid: final_tx.txid(),
+						index: if change_output_position == 0 { 1 } else { 0 },
+					};
+					// Give the funding transaction back to LDK for opening the channel.
+					loop_channel_manager
+						.funding_transaction_generated(&temporary_channel_id, outpoint);
+					pending_txs.insert(outpoint, final_tx);
+				}
+				Event::FundingBroadcastSafe { funding_txo, .. } => {
+					let funding_tx = pending_txs.remove(&funding_txo).unwrap();
+					bitcoind_client.broadcast_transaction(&funding_tx);
+					println!("\nEVENT: broadcasted funding transaction");
+					print!("> ");
+					io::stdout().flush().unwrap();
+				}
+				Event::PaymentReceived { payment_hash, payment_secret, amt: amt_msat } => {
+					let mut payments = payment_storage.lock().unwrap();
+					if let Some((Some(preimage), _, _, _)) = payments.get(&payment_hash) {
+						assert!(loop_channel_manager.claim_funds(
```

Comment:
Hmmm, so this absolutely should not be called with the info from `PaymentReceived` - it breaks our privacy by exposing us to attacks where a middle-hop node can figure out if a payment is ours, and also allows a sender to send us less than the invoice amount. The first I think we can solve at the API level by removing the payment_secret from the PaymentReceived event - users should never really need to check it themselves and should just be storing it and fetching it to provide to claim_funds. The second we could do something similar, but it feels a bit more....bleh?

Reply:
Hum so if we get a `PaymentReceived` event but can't find the payment info, then we need to call `channel_manager.fail_htlc_backwards(..)` which is parameterized by an optional PaymentSecret. So, I guess we can leave it as `None` in that case?

Reply:
Hmmm, damn, yea, so in that case it *does* need to match. I guess it really needs to be `if exists && payment_secret_matches { claim } else { fail(with_event_data) }`